File: Shared/Services/UC/BaseUCService.cs
using System.Net.Http.Json;

using Shared.Requests;
using Shared.Responses;
using Shared.Utilities;

namespace Shared.Services.UC
{
public abstract class BaseUCService<TRequest, TResponse>
    where TRequest : BaseRequest
    where TResponse : BaseResponse
{
    private readonly HttpClient _client = new();

#if DEBUG
    private const string _URLBase = "http://localhost:32406/";
#else
    private const string _URLBase = "http://162.55.32.18:80/";
#endif

    protected abstract string GetControllerName();

    protected enum EDBAction
    {
        FromTo,
        Convert
    }

    protected enum EHTTPRequest
    {
        Post,
        Get
    }

    public virtual async Task<FromToResponse> FromTo() => await Request<FromToResponse>(EHTTPRequest.Get,
                                                                                        EDBAction.FromTo.ToString());
    public virtual async Task<TResponse> Convert(TRequest request) => await Request<TResponse>(
        EHTTPRequest.Post, EDBAction.Convert.ToString(), request);

    protected async Task<Response> Request<Response>(EHTTPRequest requestHTTP, string action, object? value = null)
    {
        var uri = $"{_URLBase}{GetControllerName()}/{action}";

        HttpResponseMessage? response = requestHTTP switch {
            EHTTPRequest.Get => await _client.GetAsJsonAsync(uri, value),
            EHTTPRequest.Post => await _client.PostAsJsonAsync(uri, value),
            _ => throw new ArgumentException($"The EHTTPRequest type '{requestHTTP}' is not allowed!"),
        };

        return await ProcessResponse<Response>(response);
    }

    private static async Task<Response> ProcessResponse<Response>(HttpResponseMessage message)
    {
        var parseException = new ArgumentException("Could not ReadFromJsonAsync the message content!");
        if (message.IsSuccessStatusCode)
        {
            return await message.Content.ReadFromJsonAsync<Response>() ?? throw parseException;
        }
        else
        {
            var error = await message.Content.ReadFromJsonAsync<ErrorResponse>() ?? throw parseException;

            var exceptionName = $"Shared.Exceptions.{error.TypeException}Exception";
            throw (Exception)Activator.CreateInstance(Type.GetType(exceptionName)!, error)!;
        }
    }
}
}

File: Web/src/services/uc/BaseUCService.ts
import axios, { AxiosRequestConfig, AxiosResponse } from "axios";
import { BaseRequest } from "../../models/requests/BaseRequest";
import { BaseResponse } from "../../models/responses/BaseResponse";
import { ErrorResponse } from "../../models/responses/ErrorResponse";
import { CreateResponse } from "../../models/responses/ResponseExtensions";
import { CreateException } from "../../models/exceptions/ExceptionExtensions";

enum EDBAction {
  FromTo = "FromTo",
  Convert = "Convert",
}

export enum EHTTPRequest {
  Post,
  Get,
}

export class BaseUCService<
  TRequest extends BaseRequest,
  TResponse extends BaseResponse
> {
  readonly urlBase: string = "";
  readonly axiosConfig = {
    validateStatus: () => true,
  };

  protected constructor() {
    if (import.meta.env.DEV) {
      this.urlBase = "http://localhost:32406/";
    } else {
      this.urlBase = "http://162.55.32.18:32406/";
    }
  }

  protected GetControllerName() {
    throw new Error("Not implemented");
  }

  public async FromTo() {
    return await this.Request(EHTTPRequest.Get, EDBAction.FromTo.toString());
  }

  public async Convert(request: TRequest) {
    return await this.Request(
      EHTTPRequest.Post,
      EDBAction.Convert.toString(),
      request
    );
  }

  protected async Request(
    requestHTTP: EHTTPRequest,
    action: string,
    value?: TRequest
  ) {
    const uri = `${this.urlBase}${this.GetControllerName()}/${action}`;

    let promise;
    switch (requestHTTP) {
      case EHTTPRequest.Get:
        promise = axios.get<TResponse>(uri, this.axiosConfig);
        break;

      case EHTTPRequest.Post:
        promise = axios.post<TResponse>(uri, value, this.axiosConfig);
        break;

      default:
        throw new Error(
          `The EHTTPRequest type '${requestHTTP}' is not allowed!`
        );
    }

    return this.ProcessResponse(await promise);
  }

  private async ProcessResponse(
    result: AxiosResponse<TResponse, AxiosRequestConfig<any>>
  ) {
    const response = CreateResponse(result.data.type, result.data);
    if (result.status === 200) {
      return response as TResponse;
    } else {
      const error = response as ErrorResponse;
      throw CreateException(error.typeException, error);
    }
  }
}
